engine/src/main/java/org/terasology/physics/engine/VoxelFluidWorldSystem.java

```diff
@@ -0,0 +1,157 @@
+import org.terasology.world.WorldProvider;
+import org.terasology.world.block.Block;
+import org.terasology.world.block.BlockComponent;
+import org.terasology.world.chunks.Chunk;
+import org.terasology.world.chunks.ChunkConstants;
+import org.terasology.world.chunks.ChunkProvider;
+import org.terasology.world.chunks.event.BeforeChunkUnload;
+import org.terasology.world.chunks.event.OnChunkLoaded;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+import static org.terasology.physics.bullet.BulletPhysics.AABB_SIZE;
+
+/**
+ * Manages voxel shape for fluid and updates collision state between Bullet and Terasology
+ */
+@RegisterSystem
+public class VoxelFluidWorldSystem extends BaseComponentSystem {
+    @In
+    private PhysicsEngine physics;
+    @In
+    private EntityManager entityManager;
+    @In
+    private WorldProvider worldProvider;
+    @In
+    private ChunkProvider chunkProvider;
+
+    private final TShortHashSet registered = new TShortHashSet();
+
+    private btRigidBody.btRigidBodyConstructionInfo blockConsInf;
+    private btVoxelShape worldShape;
+    private VoxelCollisionAlgorithmWrapper wrapper;
+    private btRigidBody rigidBody;
+
+    @Override
+    public void initialise() {
+        if (physics instanceof BulletPhysics) {
+            btDiscreteDynamicsWorld discreteDynamicsWorld = ((BulletPhysics) physics).getDiscreteDynamicsWorld();
+
+            wrapper = new VoxelCollisionAlgorithmWrapper(ChunkConstants.SIZE_X, ChunkConstants.SIZE_Y,
+                ChunkConstants.SIZE_Z);
+            worldShape = new btVoxelShape(wrapper, new Vector3f(-AABB_SIZE, -AABB_SIZE, -AABB_SIZE),
+                new Vector3f(AABB_SIZE, AABB_SIZE, AABB_SIZE));
+
+            Matrix4f matrix4f = new Matrix4f();
+            btDefaultMotionState blockMotionState = new btDefaultMotionState(matrix4f);
+
+            blockConsInf = new btRigidBody.btRigidBodyConstructionInfo(0, blockMotionState, worldShape, new Vector3f());
+            rigidBody = new btRigidBody(blockConsInf);
+            rigidBody.setCollisionFlags(btCollisionObject.CollisionFlags.CF_STATIC_OBJECT | rigidBody.getCollisionFlags()); // voxel world is added to static collision flag
+            short mask = (short) StandardCollisionGroup.LIQUID.getFlag(); // interacts with anything but static and
+            // liquid
+            discreteDynamicsWorld.addRigidBody(rigidBody, physics.combineGroups(StandardCollisionGroup.WORLD), mask); // adds rigid body to world
+        }
+
+        super.initialise();
+    }
+
+    /**
+     * update voxel info for the wrapper for the associated block id
+     *
+     * @param block the block
+     */
+    private void tryRegister(Block block) {
+        short id = block.getId();
+        if (!registered.contains(id)) {
+            if (block.isLiquid()) {
+                btCollisionShape shape = ((BulletCollisionShape) block.getCollisionShape()).underlyingShape;
+                btVoxelInfo info = new btVoxelInfo(false,
```

Comment:
This first argument specifies whether the block should be raytracable. This has to be set to true, otherwise the liquid collider can't actually be used. I tested this PR in its current state with the liquid placement PR, and liquids can't actually be targeted with fluid containers.